Major Overhaul

dr-bonez opened the pull request:
So this is a big one. It was also a bit heavy handed with the removal of configure_me since I don't know it as well, but I tried to make it easy to sub back in. All that needs to be done on that front is fill in the create_env::cfg_me module.

With this overhaul we get:
- Updated to state of the art tokio and hyper. This is what motivated the bulk of the changes.
- Pruned nodes pretend to not be! If a block that is being asked for is already pruned, it will fetch it from any peers advertising "NETWORK". This will enable things like c-lightning against a pruned node.
- Made the behavior of the api (error codes, etc) look as identical as possible to Bitcoin Core 0.20.x

This has all been naively tested, but make sure to do a heavier battery of tests before putting this in a mission critical place.

Kixunil reviewed (changes requested):
Firstly, thanks for the PR!

I didn't have much time to review it, just focused on details that are easy to get wrong and things I noticed along the way. I will continue with the review later but so far I'm letting you know about the issues I found so that you can start working on it sooner.

One big difference that I found is this actually doesn't behave like Bitcoin Core. Bitcoin Core pre-scans all the RPC calls in batch and if any is forbidden, it fails the whole batch. This was discussed here: https://github.com/bitcoin/bitcoin/issues/19087

To maintain compatibility, it needs to behave the same.

Also I didn't check yet but please make sure this works with Rust 1.41.1. That's the minimum Rust version I have to support.

dr-bonez commented on src/client.rs:
```diff
@@ -0,0 +1,523 @@
+                    serde_json::from_slice(body.as_ref())?;
+                let body = serde_json::to_vec(
+                    &futures::stream::iter(forwarded_res)
+                        .chain(intercepted_recv)
```
I'll probably just build the output with https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.merge_by after collecting the intercepted_recv. This relies on bitcoin core always returning batched responses in the same order as the requests, though. Can we rely on that?

dr-bonez commented on src/lib.rs:
```diff
@@ -0,0 +1,36 @@
+use crate::proxy::proxy_request;
+pub use crate::users::{User, Users};
+
+pub async fn main(env: &'static Env) -> Result<(), Error> {
```
If you don't mid unsafe code, there's a solution to this without using `Arc`'s which have more overhead than static borrows. I'm fine with it either way, I don't care about either the valgrind leak report, the overhead of the Arcs, or the unsafe code required to do it without them, so whatever you prefer.

Kixunil replied:
I don't expect `Arc` to have a significant overhead. It should be cloned once for each request, which implies syscall being made. The cost of syscalls is much higher, so atomic operation on top of that should not be a huge difference.

If someone shows me data from a profiler that contradict this I'd be happy to reconsider, of course.

Just quickly answered, will review the rest later.

dr-bonez commented on src/create_env/start9.rs:
```diff
@@ -0,0 +1,163 @@
+                    .parse()?,
+                )?
+            }
+        },
```
Can do, but I'm thinking maybe this code should be pulled out of the crate and moved to our wrapper [here](https://github.com/Start9Labs/btc-rpc-proxy-wrapper) anyway, importing the proxy as a lib, especially since the config it is referencing is defined there.

Kixunil commented on src/fetch_blocks.rs:
```diff
@@ -0,0 +1,395 @@
+    Ok(match fetch_block_from_self(&*env, hash).await? {
+        Some(block) => Some(block),
+        None => {
+            info!(
```
I suggest decreasing to `debug`. Otherwise if you run `electrs` against this, you'll get a full log of useless messages.

Kixunil commented on src/fetch_blocks.rs:
```diff
@@ -0,0 +1,395 @@
+            if let Some(block) = fetch_block_from_peer_set(env.clone(), peers, hash).await {
+                Some(block)
+            } else {
+                warn!(env.logger, "Could not fetch block from peers.");
```
In this situation the proxy fails at pretending the node is not pruned, so maybe `error!` is more appropriate.

Kixunil commented on src/proxy.rs:
```diff
@@ -0,0 +1,90 @@
+                                                req.method.0
+                                            )
+                                        } else {
+                                            info!(
```
These two should be debug to avoid huge amount of messages.

dr-bonez replied:
I thought `slog` compiled out log levels of `debug` and below in release builds? I have found these log statements incredibly useful, so I'd like to not lose them in release builds. If there is a simple way to override that behavior in `slog` I'd love to know/.

Kixunil replied:
Yes, it can be enabled by feature flag. Feel free to enable it (or I can do it) by default.

Kixunil commented on src/users.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "kebab-case")]
```
Renaming this is backwards-incompatible and breaks expectations of `configure_me` behavior. Please don't rename it.

dr-bonez replied:
yeah, sorry, did that because of yaml convention. It would be nice to have serde be able to let the backend choose the casing, such that json can get `camelCase`, yaml can get `kebab-case`, etc without changing the serialize impl. I can revert that here, and move the translation to my configuration code, no reason these kinds of opinions should be buried deep in the codebase like that.

Kixunil replied:
Yeah that makes sense. Over time it occurred to me that `serde` is more suitable at the boundaries and whatever it deserializes should be converted to internal representation. It may look weird but I don't know of a better approach.

Kixunil commented on Cargo.toml:
```diff
@@ -1,39 +1,62 @@
 spec = "config_spec.toml"
 
+[features]
+compat = []
```
I suggest renaming to `old_rust` or adding a comment.
We could also check the version in build script and set the feature automatically.

Kixunil commented on src/fetch_blocks.rs:
```diff
@@ -0,0 +1,395 @@
+}
+
+#[derive(Debug)]
+pub struct PeerList {
```
I suggest renaming to `Peers` as this is not a list.

Kixunil commented on src/env.rs:
```diff
@@ -0,0 +1,40 @@
+        Arc::new(self)
+    }
+    pub async fn get_peers(&self) -> Result<Vec<PeerHandle>, Error> {
+        self.peers.lock().await.get_peers(self).await
```
This pattern is notoriously known for leading into various kinds of deadlocks. A better approach is:

```rust
struct Env {
// ...
peers: RwLock<Arc<PeerList>>,
}
```

then to get the peers:

```rust
let peers = self.peers.lock_read().clone();
```

When peers need to be updated, create a new collection from scratch and swap them:

```rust
std::mem::swap(&mut *self.peers.lock_write(), &mut new_peers);
```

In order to avoid multiple tasks from re-fetching the peer list, have a separate task that periodically calls into `bitcoind` and fills `peers`. This will also avoid clients from waiting on the operation to finish. The previous value of `peers` can be kept to attempt to reuse the allocation (`Arc::get_mut()`) in this case. To make sure some sane value exists in `peers` this operation can be forced at start with blocking.

dr-bonez replied:
since I'm using an `mpsc` to cache a TCP connection on each peer, unfortunately I need a `&mut Peer` to access it. If I switched to a `mpmc`, it would allow this, but I'm not super aware of the tradeoffs.

Kixunil reviewed (commented):
From review it looks mostly usable. I didn't find significant *blockers* except for that `kebab-case`.

I have some questions and ideas as you can see. Some of them might be reasonable to do some of them in separate steps.

I'm going to try to package the modified version (without `kebab-case`) into my [Debian Repository](https://deb.ln-ask.me) and run whole test suite against it. If it works without noticeable problems, I will merge it and try to package a version that actually uses the block fetching feature and run the test suite again and then let it sit in `experimental` for a while, attempting to use it. I'll let you know if I find any problems.

MattDHill commented on README.md:
Correct, BTCPay and LND will only fight for control over pruning if the node is pruned. Either service is capable of running against a pruned node, but you cannot run them both against the _same_ pruned node.

That is the benefit of using the proxy. BTCPay and LND will no longer fight for control over pruning because they believe they are dealing with an unpruned node. It's a huge benefit I think we should highlight.

MattDHill replied:
update pushed, including downside explantion

Kixunil replied:
> Either service is capable of running against a pruned node, but you cannot run them both against the same pruned node.

This is not true. I ran BTCPayServer with LND against the same pruned node and it worked (I ran into trouble later when LND stayed down for too long but that's unrelated). The real trouble is that running LND against a pruned node is a bad idea. `electrs` can't work against pruned node at all.

ProofOfKeags replied:
Matt is referred to when manual pruning is enabled. And this is a problem for c-lightning, not lnd iirc.

Kixunil replied:
What?! c-lightning requires a pruned node that is different from BTCPay? How it's the check even implemented? And who the hell thought that such a design is a good idea?

Kixunil commented on README.md:
```diff
@@ -14,9 +14,11 @@ There's another interesting advantage: since this is written in Rust, it might s
+By connecting to your pruned Bitcoin node through Bitcoin Proxy, your node will now behave as though it is not pruned. If a user or application requires a block that is not retained by your pruned node, Bitcoin Proxy will dynamically fetch the block over the P2P network, then verify its hash against your node to ensure validity.
 
-The usefulness of this new technology is massive. It means that you can run multiple services against your pruned Bitcoin node — such as Lightning and BTCPay — without them fighting for control over the pruning. Both are happy because both believe they are dealing with an unpruned node.
+This means that you can run multiple services against your _pruned_ Bitcoin node — such as Lightning and BTCPay — without them fighting for control over the pruning. Both are happy because both believe they are dealing with an _unpruned_ node.
```
Maybe better "This means that you can run various services that would otherwise require _unpruned_ node against your _pruned_ Bitcoin node — such as Lightning and `electrs`. They are happy because they believe they are dealing with an _unpruned_ node."

Rationale:

* BTCPaySever doesn't need unpruned node anyway
* This has nothing to do with using more than one service
* "that would otherwise require" sounds more understandable to me than "fighting for control"

BTW good point about bandwidth below. :)

dr-bonez commented on src/client.rs:
```diff
@@ -0,0 +1,374 @@
+        }
+    }
+}
+impl<'de> Deserialize<'de> for SingleOrBatchRpcRequest {
```
Untagged simply tries deserializing each in order, failing over. Also, if none of the variants match, the error message is useless. This method is more code but it's more efficient, since it decides which variant to deserialize into based on whether it's an object or array.

Kixunil replied:
Makes sense

dr-bonez commented on src/create_env.rs:
```diff
@@ -0,0 +1,49 @@
+
+include_config!();
+
+pub async fn create_env() -> Result<Env, Error> {
```
It's a vestige of an old way I was doing this. In my use case, I have different code that creates `Env`, which was easier to do async. I had both files follow the same api, and could be switched between with a feature flag. But then I realized it was better to just export the proxy as a lib so I didn't have to put Embassy specific code in the project.

Kixunil replied:
So could it be removed? I guess it could help compile times and reviewers.

dr-bonez commented on src/env.rs:
```diff
@@ -0,0 +1,40 @@
+use std::net::SocketAddr;
```
It isn't just config, it holds global application state which is derived from config, so it didn't feel right to call it `Config`. I'd be fine with `State` which I think is what actix does for this pattern. Many other projects I've worked on call it `Context`, which would be fine too.

Kixunil replied:
`State` is a good name for what you described.

dr-bonez commented on src/fetch_blocks.rs:
```diff
@@ -0,0 +1,395 @@
+    }
+}
+
+async fn fetch_block_from_peer<'a>(
```
the idea was that if an error (or panic) occurred before the connection was used, it would still be reusable, but in this scenario, we don't really use that, so this pattern isn't really useful.

Kixunil replied:
OK, I will look at it after merging and try to simplify it. Seems better than adding it into this already huge PR.

dr-bonez commented on src/client.rs:
```diff
@@ -0,0 +1,374 @@
+}
+
+#[derive(Debug)]
+pub struct RpcClient {
```
it doesn't perfectly fit the use case. I could replace `call` with that, but the proxying code implemented here (that returns a hyper::Response) would need to be written anyway.

Kixunil replied:
Ah, right. OK.

Kixunil commented on src/util.rs:
```diff
@@ -0,0 +1,91 @@
+            }
+        }
+        fn strip_suffix<'a>(&'a self, suffix: &str) -> Option<&'a str> {
+            if let Some(s) = self.rmatches(prefix).next() {
```
`s/prefix/suffix/`

Kixunil commented on src/rpc_methods.rs:
```diff
@@ -0,0 +1,371 @@
+
+use crate::client::RpcMethod;
+use crate::util::{Either, HexBytes};
+
```
Missing:

```rust
#[cfg(feature = "compat")]
use crate::util::compat::StrCompat;
```

Kixunil commented on src/lib.rs:
```diff
@@ -0,0 +1,44 @@
+#[macro_use]
```
We probably need `#![cfg_attr(feature = "compat"), allow(unstable_name_collisions)]`

Kixunil commented on src/create_env.rs:
```diff
@@ -0,0 +1,49 @@
+
+use config::ResultExt;
+
+include_config!();
```
This macro is currently screwed up, we need to expand it manually for the time being:

```rust
mod config {
    include!(concat!(env!("OUT_DIR"), "/configure_me_config.rs"));
}
use self::config::{Config, ResultExt};
```

Kixunil commented on src/create_env.rs:
```diff
@@ -0,0 +1,49 @@
+use slog::Drain;
+use tokio::sync::Mutex;
+
+use config::ResultExt;
```
This needs to be removed because of the one below.

Kixunil commented on src/main.rs:
```diff
@@ -1,331 +1,15 @@
-extern crate slog;
-extern crate slog_term;
-extern crate slog_async;
 #[macro_use]
```
`#![type_length_limit="1630928"]` needs to be added to the top.

Kixunil commented on Cargo.toml:
```diff
@@ -1,39 +1,62 @@
+path = "src/lib.rs"
+
+[[bin]]
+name = "btc-rpc-proxy"
```
Name of the binary is a part of API, so it should be `snake_case`

Kixunil commented on src/create_env.rs:
```diff
@@ -0,0 +1,49 @@
+include_config!();
+
+pub async fn create_env() -> Result<Env, Error> {
+    let (config, _) = config::Config::including_optional_config_files(std::iter::empty::<&str>())
```
Remove `config::` to avoid a warning and keep the code a bit cleaner.

Kixunil reviewed (changes requested):
So I finally found the time to actually try it and I found a few issues that prevent compilation/usage.

The tests just started running, so I'll let you know when they finish.

Kixunil commented:
One more thing you might want to know: I plan to add systemd support via my new crate [`systemd_socket`](https://github.com/Kixunil/systemd_socket) after merging this. Shouldn't change much for you, but let me know if there's some issue I need to address. The crate is not published yet, I'm going to publish it when I test it with a real-world application.

dr-bonez commented:
Ok, I think I've addressed everything.

Also, we had an issue with some customers blowing through their data caps during lnd/c-lightning IBD, due to the fact that we reach out to all peers concurrently for the same block. I have added a config option that can limit this concurrency.

As a thought for later, perhaps 0.3.0, I have been thinking that there may be a lot of different features that the proxy may want to provide besides this dynamic block fetching, and perhaps the best way to do it would be with a plugin system a-la c-lightning. This way, plugins could register to intercept rpc calls, and provide additional services. In this scenario, this block fetching would be re-implemented as a plugin which is registered to intercept the `getblock` rpc command.

Kixunil commented:
I was thinking very simmilar things long before. :) I would definitely want to also provide `txindex` via `electrs` (I added the required method to `electrs` recently) but I think such change is too small for running yet another service.

However, what I'd really love to see is simulated wallet. This could be a separate daemon. Yet, a simpler implementation may be that the multiple wallets feature of bitcoind is used and the proxy merely makes sure the caller doesn't attempt to access other wallets (and if it doesn't specify which wallet, the proxy forces the right one). Another interesting possibility is to redirect wallet calls to NBXplorer, so that the same wallet can be used for PayJoin and Eclair.

Kixunil commented on src/users.rs:
```diff
@@ -0,0 +1,172 @@
+pub struct User {
+    pub password: String,
+    pub allowed_calls: HashSet<String>,
+    pub fetch_blocks: bool,
```
This needs `#[serde(default)]` to maintain backwards compatibility.

Kixunil commented:
I attempted to use it with btcpayserver on regtest and I'm getting this error:

```
INFO public called getblockchaininfo: ERROR -1 EOF while parsing a value at line 1 column 0
```

Kixunil commented:
Also `#![type_length_limit="1717928"]` [is still missing](https://github.com/cryptoadvance/specter-desktop/issues/722#issuecomment-735797737)

Kixunil commented on src/client.rs:
```diff
@@ -0,0 +1,371 @@
+impl RpcClient {
+    pub fn new(auth: AuthSource, uri: Uri) -> Result<Self, Error> {
+        Ok(RpcClient {
+            authorization: auth.try_load()?,
```
This is wrong, `bitcoind` may restart while proxy is running and it'll re-generate a new cookie. Indeed it breaks my deployment. The previous code was intentionally loading the cookie whenever it was needed.

Optimization caching the cookie and reloading&retrying on error is allowed and encouraged.

dr-bonez replied:
instead of retrying on error, I'm just to invalidating the cache based on the modified timestamp of the cookie file. Hope that works.

Kixunil replied:
Hmm, that works although maybe it has smaller performance saving over just reading those few bytes. I guess you're saving at least one syscall (reading is open+read or seek+read; stat is just one syscall).

Kixunil reviewed (changes requested):
Found another blocker.

dr-bonez commented:
> EOF while parsing a value

This seems to indicate that bitcoin core is returning a response that isn't valid for JSONRPC. I have noticed this on 401s, but I'm sure other things can cause this. I can make the error message better.

dr-bonez commented:
> Also `#![type_length_limit="1717928"]` [is still missing](https://github.com/cryptoadvance/specter-desktop/issues/722#issuecomment-735797737)

wasn't missing, just isn't high enough :)
it seems to depend on rustc version, so maybe I'll give it some margin.

Kixunil commented:
> This seems to indicate that bitcoin core is returning a response that isn't valid for JSONRPC. I have noticed this on 401s, but I'm sure other things can cause this. I can make the error message better.

Yeah, it was due to the cookie. Would be nice to improve it a bit.

> just isn't high enough :)

Hmm, interesting, giving it some margin is a good idea.

Kixunil commented:
I'm going to test again tomorrow. Any thoughts on solving rescan?

dr-bonez commented:
unfortunately I think rescan is too internal to bitcoind to mock out individually. We would basically have to rebuild the wallet api from scratch to handle this.

Kixunil commented:
Damn, that sucks. Cc @openoms

Kixunil commented:
Oh, I think I know how it could be achieved:

Either use `electrs` or scan the chain one block at a time to find transactions related to the addresses from the wallet. Each time a transaction is found call `importprunedfunds` to notify `bitcoind` of its existence.

I guess it'd be painfully slow without `electrs` and eat ton of bandwidth, so requiring electrs for starters would make more sense IMO. Alternatively try to use block filters, but I don't know if block filters can work with pruning.

Kixunil commented:
I still get errors when attempting to connect. I believe most likely because cookie is not sent base64 encoded.

Dump from Wireshark (don't worry about cookie, it's a disposable regtest node)

```
POST / HTTP/1.1
authorization: Basic __cookie__:958d85f51f2851e42a9edd57a1a633418a146643ed1b796d6f6d9c0ad8bfe548
host: 127.0.0.1:18442
content-length: 60

{"id":"1606826395898","method":"getnetworkinfo","params":[]}HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="jsonrpc"
Date: Tue, 01 Dec 2020 12:39:56 GMT
Content-Length: 0
Content-Type: text/html; charset=ISO-8859-1

POST / HTTP/1.1
authorization: Basic __cookie__:958d85f51f2851e42a9edd57a1a633418a146643ed1b796d6f6d9c0ad8bfe548
host: 127.0.0.1:18442
content-length: 60

{"id":"1606826425876","method":"getnetworkinfo","params":[]}HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="jsonrpc"
Date: Tue, 01 Dec 2020 12:40:26 GMT
Content-Length: 0
Content-Type: text/html; charset=ISO-8859-1
```

The content matches the cookie file, but the encoding is not base64. Stream sent from the client to the proxy:

```
POST / HTTP/1.1
host: 127.0.0.1:18443
authorization: Basic cHVibGljOnB1YmxpYw==
content-length: 60
Connection: close

{"id":"1606826395898","method":"getnetworkinfo","params":[]}HTTP/1.1 401 Unauthorized
www-authenticate: Basic realm="jsonrpc"
date: Tue, 01 Dec 2020 12:39:56 GMT
content-type: text/html; charset=ISO-8859-1
content-length: 0
```

Kixunil commented:
Seems to work, merging and will test it in experimental for a while.
I think I'll make a few more changes to the proxy before activating this feature in my repository.
Do you want to get pinged about them?

dr-bonez commented:
yes please
